### ISSUE TITLE
Performance issue: Success for sending/receiving POST and REACT requests is inconsistent

### ISSUE THREAD

**Issue by Sophietje** [TNO/knowledge-engine#282]
In GitLab by @remi.roche.tiko.energy on Aug 19, 2021, 11:33

We noticed that we cannot get consistent results when testing a POST/REACT exchange, both locally and with a partner.

Here is what happened (on 18.08.2021 between 9:50 and 10:40 Paris time):
- We tested with a partner a POST/REACT exchange. We were sending the POST and our partner the REACT.
- Out of a dozen tries, only a few POST were received by our Partner’s react.
- We received the feedback from REACT only once.
- Then we tried with our own react script, which received at once a couple of POST requests (maybe "stuck" somewhere ?). At the same time our partner also received a POST (only one), which was not happening before. But again no answer received on POST side.
- Thinking we were doing something wrong, our partner tried with postman to send a POST and it was also not working.
- We also tried restarting everything and had troubles at the registering knowledge base step which took a long time.

So to summarize:
- we can send a POST request
- the REACT sides don't receive 100% of the time (more like 25%), whether it is locally running or from our partner's platform
- When receiving the POST, the REACT side sends the confirmation properly
- The POST side rarely receives the confirmation from REACT (around 5% of the time)

I saw some similar tickets about slow performance and not receiving the response of the react https://gitlab.inesctec.pt/interconnect/knowledge-engine/-/issues/250

We tried both right after registering, or after waiting a bit. It does not change anything.
In this state, we cannot test properly our connections with our partners.

**Comment by Sophietje** [comment]
In GitLab by @barry.nouwt.tno.nl on Aug 19, 2021, 14:55

Hi @remi.roche.tiko.energy, thanks for your report! This is indeed strange behaviour that prevents your testing. 

The only possible explanation that I can think of for this behaviour (apart from a bug) is that there are other Knowledge Bases (than the one of your partner) with matching Knowledge Interactions. *If* that is the case, the Knowledge Engine will contact one Knowledge Base after the other (so not in parallel) and the ordering is random. This would also explain not getting a confirmation from the REACT (because the Knowledge Engine waits until all Knowledge Bases have responded).

A few questions for additional info:
- what are the Knowledge-Base-Ids of the KBs involved (this makes checking our logs a bit easier)?
- what Graph Patterns were you using (so we can check if other unexpected KBs also use the same ones)?

Regards, Barry

**Comment by Sophietje** [comment]
In GitLab by @remi.roche.tiko.energy on Aug 20, 2021, 06:09

Hello Barry, 

Thanks for your help. We did a lot of cleaning after some tests so the knowledgebase ids may have change:
POST KB Id: https://ke.interconnectproject.eu/adapter/ebe5e74a0cec9c629aab75ae8127f1dfff152da2c8e3338065a59a839e56a6fc986d6dd89940ed20dba920dbd896f806f80079372e646e254fff24d834a78c10
local REACT KB Id: https://ke.interconnectproject.eu/adapter/9a04ecfa09c873c18bb155546dec2f0c40d492c1f415e4f1b6a750a7a458aa8b7c7079c9d33e2c58e29210dbe40e7d599c9846b440913d3e7e953a4d41418f71
partner's REACT KB Id: waiting for feedback from partner

And you will find attached below the gp pattern we used.

[tiko_device_registration.gp](https://gitlab.inesctec.pt/interconnect/knowledge-engine/uploads/86a4ad713c2e459f3e04dbc941b09fb9/tiko_device_registration.gp)

We can also redo some tests at a specific time or during a call if you want, and share the code.

**Comment by Sophietje** [comment]
In GitLab by @michel.hours.inetum.world on Aug 20, 2021, 06:52

Hello Barry and Remi,

Here is the Knowledge-Base-Id of the REACT side:

https://ke.interconnectproject.eu/adapter/96386ede738dddfba774b76b4a189489dda1ac923a11b45b93a61493c6dedceca2fb925ebe79ae41323df709eed0cb6289f01e3d8029761532060039156213ec

The french-house-data-definition service is using the graph patterns below:

graphpattern.houseregistration=?building <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <https://saref.etsi.org/saref4bldg/Building> . ?building <http://ontology.tno.nl/building#LocatedIn> ?spatialThing . ?spatialThing <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.w3.org/2003/01/geo/wgs84_pos#SpatialThing> . ?spatialThing <http://www.geonames.org/ontology#postalCode> ?zipCode . ?building <http://ontology.tno.nl/building#hasEnergyClass> ?energyClass . ?building <http://ontology.tno.nl/building#energyProvider> ?energyProvider . ?building <http://ontology.tno.nl/building#flexibilityManager> ?flexibilityManager . ?building <http://ontology.tno.nl/building#communityID> ?communityID . ?building <https://saref.etsi.org/saref4bldg/hasSpace> ?buildingSpace . ?buildingSpace <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <https://saref.etsi.org/saref4bldg/BuildingSpace> . ?buildingSpace <https://saref.etsi.org/saref4bldg/contains> ?buildingDevice . ?buildingDevice <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <https://saref.etsi.org/saref4bldg/BuildingDevice> . ?buildingDevice <http://tno.io/PowerLimit#hasContractualPowerLimit> ?powerSubscribed

graphpattern.equipmentregistration=?building <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <https://saref.etsi.org/saref4bldg/Building> . ?building <https://saref.etsi.org/saref4bldg/hasSpace> ?buildingSpace . ?buildingSpace <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <https://saref.etsi.org/saref4bldg/BuildingSpace> . ?buildingSpace <https://saref.etsi.org/saref4bldg/contains> ?device . ?device <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <https://saref.etsi.org/core/Device> . ?device <https://www.example.org/interconnect/hasClassification> ?classification . ?device <http://www.w3.org/2000/01/rdf-schema#comment> ?description . ?device <https://www.example.org/interconnect/hasServiceProviderEMSOwner> ?serviceProviderEMSOwner

graphpattern.trafficlightstate=?building <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <https://saref.etsi.org/saref4bldg/Building> . ?building <https://saref.etsi.org/saref4bldg/hasSpace> ?buildingSpace . ?buildingSpace <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <https://saref.etsi.org/saref4bldg/BuildingSpace> . ?buildingSpace <https://saref.etsi.org/saref4bldg/contains> ?device . ?device <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <https://saref.etsi.org/core/Device> . ?device <https://saref.etsi.org/core/makesMeasurement> ?measurement . ?measurement <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <https://saref.etsi.org/core/Measurement> . ?measurement <https://saref.etsi.org/core/hasValue> ?trafficLightState . ?measurement <https://saref.etsi.org/core/hasTimestamp> ?timeStamp

graphpattern.dynamictariffcollection=?building <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <https://saref.etsi.org/saref4bldg/Building> . ?building <https://www.example.org/interconnect/hasForecast> ?fc . ?fc <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://ontology.tno.nl/interconnect/forecast#forecast> . ?fc <http://ontology.tno.nl/interconnect/forecast#isExpressedAs> ?ts . ?ts <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://ontology.tno.nl/interconnect/forecast#timeSeries> . ?ts <http://ontology.tno.nl/interconnect/forecast#hasSeriesOfDataPoint> ?dp . ?dp <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://ontology.tno.nl/interconnect/forecast#dataPoint> . ?dp <http://ontology.tno.nl/interconnect/forecast#hasTimeStamp> ?timestamp . ?timestamp <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.w3.org/2006/time#Interval> . ?timestamp <http://www.w3.org/2006/time#hasBeginning> ?beginning . ?beginning <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.w3.org/2006/time#Instant> . ?beginning <http://www.w3.org/2006/time#inXSDDateTimeStamp> ?beginDateTime . ?timestamp <http://www.w3.org/2006/time#hasEnd> ?end . ?end <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.w3.org/2006/time#Instant> . ?end <http://www.w3.org/2006/time#inXSDDateTimeStamp> ?endDateTime . ?dp <http://ontology.tno.nl/interconnect/forecast#hasQuantity> ?q . ?q <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.ontology-of-units-of-measure.org/resource/om-2/Quantity> . ?q <http://www.ontology-of-units-of-measure.org/resource/om-2/hasValue> ?value . ?value <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.ontology-of-units-of-measure.org/resource/om-2/Measure> . ?value <http://www.ontology-of-units-of-measure.org/resource/om-2/hasNumericalValue> ?costValue . ?value <http://www.ontology-of-units-of-measure.org/resource/om-2/hasUnit> ?unit

graphpattern.engieequipmentloadcurves=?building <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <https://saref.etsi.org/saref4bldg/Building> . ?building <https://saref.etsi.org/saref4bldg/hasSpace> ?buildingSpace . ?buildingSpace <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <https://saref.etsi.org/saref4bldg/BuildingSpace> . ?buildingSpace <https://saref.etsi.org/saref4bldg/contains> ?device . ?device <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <https://saref.etsi.org/core/Device> . ?device <https://www.example.org/interconnect/hasTimeseries> ?ts . ?ts <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://ontology.tno.nl/interconnect/forecast#timeSeries> . ?ts <http://ontology.tno.nl/interconnect/forecast#hasSeriesOfDataPoint> ?dp . ?dp <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://ontology.tno.nl/interconnect/forecast#dataPoint> . ?dp <http://ontology.tno.nl/interconnect/forecast#hasTimeStamp> ?timestamp . ?timestamp <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.w3.org/2006/time#Instant> . ?timestamp <http://www.w3.org/2006/time#inXSDDateTimeStamp> ?dateTime . ?dp <http://ontology.tno.nl/interconnect/forecast#hasQuantity> ?q . ?q <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.ontology-of-units-of-measure.org/resource/om-2/Quantity> . ?q <http://www.ontology-of-units-of-measure.org/resource/om-2/hasValue> ?value . ?value <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.ontology-of-units-of-measure.org/resource/om-2/Measure> . ?value <http://www.ontology-of-units-of-measure.org/resource/om-2/hasNumericalValue> ?load . ?value <http://www.ontology-of-units-of-measure.org/resource/om-2/hasUnit> ?unit

graphpattern.engieinitialtemperaturescheduling=?building <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <https://saref.etsi.org/saref4bldg/Building> . ?building <https://saref.etsi.org/saref4bldg/hasSpace> ?buildingSpace . ?buildingSpace <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <https://saref.etsi.org/saref4bldg/BuildingSpace> . ?buildingSpace <https://saref.etsi.org/saref4bldg/contains> ?device . ?device <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <https://saref.etsi.org/core/Device> . ?device <https://www.example.org/interconnect/hasForecast> ?fc . ?fc <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://ontology.tno.nl/interconnect/forecast#forecast> . ?fc <http://ontology.tno.nl/interconnect/forecast#isExpressedAs> ?ts . ?ts <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://ontology.tno.nl/interconnect/forecast#timeSeries> . ?ts <http://ontology.tno.nl/interconnect/forecast#hasSeriesOfDataPoint> ?dp . ?dp <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://ontology.tno.nl/interconnect/forecast#dataPoint> . ?dp <http://ontology.tno.nl/interconnect/forecast#hasTimeStamp> ?timestamp . ?timestamp <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.w3.org/2006/time#Instant> . ?timestamp <http://www.w3.org/2006/time#inXSDDateTimeStamp> ?dateTime . ?dp <http://ontology.tno.nl/interconnect/forecast#hasQuantity> ?q . ?q <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.ontology-of-units-of-measure.org/resource/om-2/Quantity> . ?q <http://www.ontology-of-units-of-measure.org/resource/om-2/hasValue> ?value . ?value <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.ontology-of-units-of-measure.org/resource/om-2/Measure> . ?value <http://www.ontology-of-units-of-measure.org/resource/om-2/hasNumericalValue> ?temperature . ?value <http://www.ontology-of-units-of-measure.org/resource/om-2/hasUnit> ?unit

graphpattern.housedeletion=?building <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <https://saref.etsi.org/saref4bldg/Building>

graphpattern.equipmentdeletion=?building <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <https://saref.etsi.org/saref4bldg/Building> . ?building <https://saref.etsi.org/saref4bldg/hasSpace> ?buildingSpace . ?buildingSpace <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <https://saref.etsi.org/saref4bldg/BuildingSpace> . ?buildingSpace <https://saref.etsi.org/saref4bldg/contains> ?device . ?device <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <https://saref.etsi.org/core/Device>

**Comment by Sophietje** [comment]
In GitLab by @barry.nouwt.tno.nl on Aug 30, 2021, 06:44

Hey everyone, 

I am still not exactly sure what has caused this, but what I said above is still my guess:

> The only possible explanation that I can think of for this behaviour (apart from a bug) is that there are other Knowledge Bases (than the one of your partner) with matching Knowledge Interactions.

So, a few additional remarks to keep in mind when testing again:
- I noticed multiple Knowledge Bases with similar names like: kb_tiko_ask1, kb_tiko_ask2, kb_tiko_ask3, etc. _What is the purpose of these Knowledge Bases?_ If you create new Knowledge Bases for new tests (without removing the old ones), you might run in the behavior where there are multiple ANSWER Knowledge Interactions that match an ASK, but those other Knowledge Bases are no longer in use and do not respond to requests from the Knowledge Engine and thus block the process indefinitely. So, before testing, make sure to:
  -  unregister all Knowledge Bases from the Knowledge Engine that you no longer use ([here](https://ke.interconnectproject.eu/rest/sc) you can see an overview of all currently registered Knowledge Bases).
  - when your service starts, please check if its Knowledge Base is already registered (with the correct Knowledge Interactions). If so, do not create a new one. Note that the Generic Adapter might automatically do this for you.

If you made sure there are no other Knowledge Bases matching, but you still see the above behavior, I would like to propose to plan a session together so I can take a look and maybe find out what is going wrong.

Kind regards,

Barry

**Comment by Sophietje** [comment]
In GitLab by @remi.roche.tiko.energy on Sep 1, 2021, 13:38

Hello Barry,

Thanks a lot for your response. We did some more tests and it took us some time to figure out if we did something wrong.
But it seems that there is indeed a problem with the platform.
When we test locally with the above GP with one variable removed, it works (post send, react answers, post gets answer)
When we test locally with the GP as above and used also by Inetum, it does not work (post sends, react never gets the post, post never gets an answer either from the local react or from Inetum's)
When we test locally with the above GP with one variable more, it works (post send, react answers, post gets answer)

We did all our tests by properly cleaning all KB and KI everytime.

So it seems that when several react listen to the same GP, it creates issues.
But also, when we send the post without the local react, we don't get anything from Inetum.

So maybe we could do a session all together with Inetum where we do tests.

**Comment by Sophietje** [comment]
In GitLab by @barry.nouwt.tno.nl on Sep 2, 2021, 11:26

Hi @remi.roche.tiko.energy, thanks for your thorough testing and reporting back! I agree that we need to plan a session together about this issue.  I will send you an email shortly!

**Comment by Sophietje** [comment]
In GitLab by @barry.nouwt.tno.nl on Sep 2, 2021, 11:28

@remi.roche.tiko.energy Is your email address remi.roche(at)tiko.energy?

**Comment by Sophietje** [comment]
In GitLab by @remi.roche.tiko.energy on Sep 2, 2021, 11:29

yes it is

**Comment by Sophietje** [comment]
In GitLab by @michel.hours.inetum.world on Sep 2, 2021, 11:29

Hello Barry,

Because I'm working on the react side I can attend to your upcoming session if needed.

**Comment by Sophietje** [comment]
In GitLab by @remi.roche.tiko.energy on Sep 6, 2021, 06:49

After a meeting with Barry and Michel, we clarified that it was not a bug but a misunderstanding.

We were doing tests with 2 REACT KB that had different answer graph patterns. One matched our post but not the other. This prevented the react process to answer to the post and so the post never got answers.

So to remember to work properly:
For the same argument graph pattern, ALL reacts and ALL posts need the same answer graph pattern. 

Also: the post will receive the answers from the react once they have all answered. The answers will be aggregated into one.
